webapp/src/blocks.tsx

```diff
@@ -789,68 +802,472 @@ export class Editor extends srceditor.Editor {
             },
             rtl: Util.isUserLanguageRtl()
         };
+        this.blocklyOptionsCache = blocklyOptions;
         return blocklyOptions;
     }
 
-    private getDefaultToolbox(showCategories = this.showToolboxCategories): HTMLElement {
-        return showCategories !== CategoryMode.None ?
-            pxt.blocks.getBaseToolboxDom().documentElement
-            : pxt.blocks.getBaseNoCategoryToolboxDom().documentElement;
-    }
-
-    filterToolbox(filters?: pxt.editor.ProjectFilters, showCategories = this.showToolboxCategories): Element {
-        this.filters = filters;
-        this.showToolboxCategories = showCategories;
-        return this.refreshToolbox();
-    }
-
     private refreshToolbox() {
-        if (!this.blockInfo) return undefined;
-
-        let toolbox = this.getDefaultToolbox(this.showToolboxCategories);
-        let tbAll: Element;
-
-        if (this.showToolboxCategories !== CategoryMode.All) {
-            tbAll = pxt.blocks.createToolbox(this.blockInfo, toolbox, CategoryMode.All, this.filters, this.extensions);
-        }
-        let tb = pxt.blocks.createToolbox(this.blockInfo, toolbox, this.showToolboxCategories, this.filters, this.extensions);
-        this.updateToolbox(tb, this.showToolboxCategories);
-
-        pxt.blocks.cachedSearchTb = tb;
-        pxt.blocks.cachedSearchTbAll = tbAll || tb;
-        return tb;
-    }
-
-    private updateToolbox(tb: Element, showCategories = this.showToolboxCategories, search = false) {
+        if (!this.blockInfo) return;
         // no toolbox when readonly
         if (pxt.shell.isReadOnly()) return;
 
-        pxt.debug('updating toolbox');
+        this.clearCaches();
+
+        const hasCategories = this.shouldShowCategories();
+
         const editor_ = (this.editor as any);
```

Comment:
Instead of casting to any you can declare those internal properties at the type at the top of this file. 
```typescript
declare namespace Blockly {
    declare interface Workspace {
        // This will be merged by TS with Blockly.d.ts
    }
}
```

Reply:
TS doesn't like that, something about ambient class.

Reply:
Resolved.